tests/components/tasmota/test_common.py

```diff
@@ -0,0 +1,328 @@
+        f"{DEFAULT_PREFIX}/{config[CONF_MAC]}/config",
+        json.dumps(config),
+    )
+    await hass.async_block_till_done()
+    async_fire_mqtt_message(
+        hass,
+        get_topic_tele_will(config),
+        get_state_online(config),
+    )
+
+    state = hass.states.get(f"{domain}.test")
+    assert state.state != STATE_UNAVAILABLE
+
+    mqtt_mock.connected = False
+    await hass.async_add_executor_job(mqtt_client_mock.on_disconnect, None, None, 0)
+    await hass.async_block_till_done()
+    await hass.async_block_till_done()
+    await hass.async_block_till_done()
+    state = hass.states.get(f"{domain}.test")
+    assert state.state == STATE_UNAVAILABLE
+
+    mqtt_mock.connected = True
+    await hass.async_add_executor_job(mqtt_client_mock.on_connect, None, None, None, 0)
+    await hass.async_block_till_done()
+    await hass.async_block_till_done()
+    await hass.async_block_till_done()
+    state = hass.states.get(f"{domain}.test")
+    assert state.state != STATE_UNAVAILABLE
+
+
+async def help_test_availability(
+    hass,
+    mqtt_mock,
+    domain,
+    config,
+):
+    """Test availability.
+
+    This is a test helper for the TasmotaAvailability mixin.
+    """
+    async_fire_mqtt_message(
+        hass,
+        f"{DEFAULT_PREFIX}/{config[CONF_MAC]}/config",
+        json.dumps(config),
+    )
+    await hass.async_block_till_done()
+
+    state = hass.states.get(f"{domain}.test")
+    assert state.state == STATE_UNAVAILABLE
+
+    async_fire_mqtt_message(
+        hass,
+        get_topic_tele_will(config),
+        get_state_online(config),
+    )
+
+    state = hass.states.get(f"{domain}.test")
+    assert state.state != STATE_UNAVAILABLE
+
+    async_fire_mqtt_message(
+        hass,
+        get_topic_tele_will(config),
+        get_state_offline(config),
+    )
+
+    state = hass.states.get(f"{domain}.test")
+    assert state.state == STATE_UNAVAILABLE
+
+
+async def help_test_availability_discovery_update(
+    hass,
+    mqtt_mock,
+    domain,
+    config,
+):
+    """Test update of discovered TasmotaAvailability.
+
+    This is a test helper for the TasmotaAvailability mixin.
+    """
+    # customize availability topic
+    config1 = copy.deepcopy(config)
+    config1[CONF_PREFIX][PREFIX_TELE] = "tele1"
+    config1[CONF_OFFLINE] = "offline1"
+    config1[CONF_ONLINE] = "online1"
+    config2 = copy.deepcopy(config)
+    config2[CONF_PREFIX][PREFIX_TELE] = "tele2"
+    config2[CONF_OFFLINE] = "offline2"
+    config2[CONF_ONLINE] = "online2"
+    data1 = json.dumps(config1)
+    data2 = json.dumps(config2)
+
+    availability_topic1 = get_topic_tele_will(config1)
+    availability_topic2 = get_topic_tele_will(config2)
+    assert availability_topic1 != availability_topic2
+    offline1 = get_state_offline(config1)
+    offline2 = get_state_offline(config2)
+    assert offline1 != offline2
+    online1 = get_state_online(config1)
+    online2 = get_state_online(config2)
+    assert online1 != online2
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6285789)